### PR TITLE
fix: `TypeExpression` - fix "JIT stack limit exhausted" error

### PR DESCRIPTION
**Pull request by kubawerlos**
Ref. https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7841#discussion_r1493362909

### CODE REVIEW

File: tests/DocBlock/TypeExpressionTest.php
Comment (Wirone) [inline tests/DocBlock/TypeExpressionTest.php:243]: Maybe we can involve the dash here too? Not necessarily to all keys, maybe only to odd/even?

File: src/DocBlock/TypeExpression.php
Comment (kubawerlos) [reply src/DocBlock/TypeExpression.php:62]: No idea, took that form the comment from previous PR.

File: tests/DocBlock/TypeExpressionTest.php
Comment (kubawerlos) [reply tests/DocBlock/TypeExpressionTest.php:243]: Then that test will fail 😄. Maybe it is time to consider using https://github.com/phpstan/phpdoc-parser.

File: src/DocBlock/TypeExpression.php
Comment (Wirone) [reply src/DocBlock/TypeExpression.php:62]: @mvorisek can you explain?

Comment (mvorisek) [reply src/DocBlock/TypeExpression.php:62]: it is "greedy possessive quantifier", let me test this PR later, it can/should be probably applied to more places

File: src/DocBlock/TypeExpression.php
Comment (mvorisek) [inline src/DocBlock/TypeExpression.php:56]: the `(?&constant)|(?&identifier)` is/was here correctly

and maybe the fix with `|(?&name)` is not that bad

here is phpstan parser code for "array shape key": https://github.com/phpstan/phpdoc-parser/blob/1.25.0/src/Parser/TypeParser.php#L839 - it allows basically integer, quoted strings, and identifier

we have a little different definion of identifier, thus the difference

here is an example of wrong phpdoc: https://phpstan.org/r/5effc31f-ddff-4e77-ad01-d7844d09beec - we should not parse more, so I prefer to revert this line

File: tests/DocBlock/TypeExpressionTest.php
Comment (mvorisek) [inline tests/DocBlock/TypeExpressionTest.php:238]: the "backtrack limit" was a problem when the type was not matched, so we should add also some large types that are invalid, like in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7843#discussion_r1493524794

File: src/DocBlock/TypeExpression.php
Comment (kubawerlos) [reply src/DocBlock/TypeExpression.php]: > it can/should be probably applied to more places

if you propose other places I will happily try to create test case for it

File: tests/DocBlock/TypeExpressionTest.php
Comment (kubawerlos) [reply tests/DocBlock/TypeExpressionTest.php]: Added, rebased to have 2 commits (instead of 8) when in 1st both new test cases - valid and invalid are failing.

File: src/DocBlock/TypeExpression.php
Comment (Wirone) [reply src/DocBlock/TypeExpression.php:62]: For those not familiar with greedy / possessive quantifiers, [here's the nice description](https://www.regular-expressions.info/possessive.html) 🙂. Thanks @mvorisek 🍻.

**Comment by mvorisek** [comment]
I submit a PR into this PR - https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/1

File: src/DocBlock/TypeExpression.php
Comment (mvorisek) [reply src/DocBlock/TypeExpression.php:62]: @kubawerlos please merge https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/1 into this PR to fix the #7839 in all places

**Review by Wirone — APPROVED** [review]

**Comment by mvorisek** [comment]
@Wirone this PR is waiting for https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/1 - or can you cherrypick it into this PR?

**Comment by Wirone** [comment]
@mvorisek I can't, I'm on mobile now, and I see there are merge conflicts in that PR. I will just cancel my approval for now until it's resolved.

**Comment by kubawerlos** [comment]
> @Wirone this PR is waiting for [6b7562617765726c6f73#1](https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/1) - or can you cherrypick it into this PR?

No, not really, https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/1 is adding 2 more `+` that are not covered by tests.

I'd go with this fix as is - as it has tests and fix. Then we can separately figure out what tests would cover the other two `+` - do I have to mention that adding logical changes to `src` without tests is not how we do things in this repo?

**Comment by mvorisek** [comment]
@kubawerlos we need to be on the same page first. I did that PR to fix https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7841#discussion_r1493362909 in all places. The test you made is logical, but can you explain the https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7843#discussion_r1493521309 need? What part of the regex is ambiguous that needs deep backtracking? TBH, I do not know, so from maintainer perspective, I fully agree "we need test always", but let me ask you another question - is this PR good to be merged without fixing the issue on all places?

**Comment by Wirone** [comment]
@mvorisek I believe @kubawerlos would love to "fix all places" but as far as I understand test cases in his PR don't cover `+` added by you, so either you should provide them with tests _or_ Kuba will be forced to figure out how to test it, which is not that good from his perspective. Let's figure out the best way so we can have a complete fix with proper test coverage 😊.

**Comment by mvorisek** [comment]
I know how to create some test case like @kubawerlos did, but I am not sure what exactly does make the regex ambiguous. For me, this is the deeper question I like to be answered.

Also, I am curious what `S` modifier (https://www.php.net/manual/en/reference.pcre.pattern.modifiers.php) does and if it imply any real performance improvement.

**Comment by kubawerlos** [comment]
> nor guarantee the testcase will be still valid with next PCRE release

nobody can predict the future.

I want to have all three newly added `+` covered with tests, so nobody will remove them and have all tests passing. I covered the one added here, I have not tried to handle others.

 I'm still hoping we can merge them separately to have nice proofs that are all covered by tests (maybe as a bonus we can make some refactoring for tests because having `testHugeType` and `createHugeArrayShapeType` is not perfect and definitely adding 2 more similar methods is not a way to go).

**Review by keradus — APPROVED** [review]
For state of this PR, I like what I see.


for state of eagenerss of extra `*` -> `*+` updates in more places, please go for it with separated PR.

I understand eagerness with having it covered with UTests, but I also acknowledge the trickyness of having utest showing internal stack limit crash without `+`. With that, I'm also OK to have further update to `*+` relying on current testbase.

**Comment by keradus** [comment]
@kubawerlos , would you mind to solve the git conflict? I think only you have rights to update the branch. 
auto-merge should then merge, when CI is green and no git conflict anymore 🙇🏻

**Comment by Wirone** [comment]
I can update @kubawerlos' PR but I'll let him decide if he wants to incorporate @mvorisek's PR before the rebase and merge here 🙂.

**Comment by kubawerlos** [comment]
> I can update @kubawerlos' PR but I'll let him decide if he wants to incorporate @mvorisek's PR before the rebase and merge here 🙂.

Let's do it in separate PR, especially that I have prepared next PR with bugfix that changes `*+` to `*` 😉 (not related to this pare, you'll see very soon).

**Comment by mvorisek** [comment]
@kubawerlos feel free to tag me there, I am curious about this regex optimization :)

**Pull request merged**
